### PR TITLE
Fix wrong ASN discovery on non-BGP Devices

### LINKED ISSUE #14752: invalid bgpLocalAs
### The problem

Some Devices carries a invalid `bgpLocalAs `

```
MariaDB [librenms]> select DISTINCT `hardware`,`bgpLocalAs` FROM `devices` WHERE `disabled` = 0 AND `ignore` = 0 AND bgpLocalAs < 2000;
+-----------------------------------------+------------+
| hardware                                | bgpLocalAs |
+-----------------------------------------+------------+
| Nexus N9K-C9336C-FX2                   |          1 |
| Planet 8P                              |          3 |
| Planet 8P                              |         41 |
| WS-C3650-24PS-L                        |         80 |
| PARADYNE                               |        131 |
| NULL                                    |        197 |
| Juniper ACX710 Internet Backbone Router |        508 |
| Juniper EX2300-24T Switch               |        514 |

```

### Output of ./validate.php

```txt
ibrenms@librenms1:~$ ./validate.php 
===========================================
Component | Version
--------- | -------
LibreNMS  | 22.11.0-21-g47cc169f4 (2022-12-12T09:02:49+01:00)
DB Schema | 2021_02_09_122930_migrate_to_utf8mb4 (248)
PHP       | 8.1.13
Python    | 3.9.2
Database  | MariaDB 10.5.15-MariaDB-0+deb11u1
RRDTool   | 1.7.2
SNMP      | 5.9
===========================================

[OK]    Composer Version: 2.5.0
[OK]    Dependencies up-to-date.
[OK]    Database connection successful
[OK]    Database Schema is current
[OK]    SQL Server meets minimum requirements
[OK]    lower_case_table_names is enabled
[OK]    MySQL engine is optimal
[OK]    
[OK]    Database schema correct
[OK]    MySQl and PHP time match
[OK]    Distributed Polling setting is enabled globally
[OK]    Connected to rrdcached
[OK]    Active pollers found
[OK]    Dispatcher Service is enabled
[OK]    Locks are functional
[OK]    Python wrapper cron entry is not present
[OK]    Redis is functional
[OK]    rrdtool version ok
[OK]    Connected to rrdcached
```

### What was the last working version of LibreNMS?

_No response_

### Anything in the logs that might be useful for us?

_No response_

### PR DESCRIPTION
**Pull request by Bierchermuesli**
There is a mighty error in bgp-peer discovery. Instead a snmp GetNext we should try to catch the first entry with a regular Get. If there is none, the device has no BGP enabled. 

```
'/usr/bin/snmpgetnext' '-v2c' '-c' 'COMMUNITY' '-OQUsv' '-m' 'BGP4-MIB' '-M' '/opt/librenms/mibs:/opt/librenms/mibs/cisco' 'udp:HOSTNAME:161' 'bgpLocalAs'
````

a BGP enabled Host (Cisco NCS)
```
... GetNextRequest(27)  15.2
... GetResponse(30)  15.2.0=<ourASN>
```
versus a non-BGP enabled Host (Cisco Nexus)
```
... GetNextRequest(27)  15.2
... GetResponse(32)  16.9.1.1.1.1=1
```
or old Brocade Switch
```
... GetNextRequest(27)  15.2
... GetResponse(32)  16.1.1.1.1.1=1
```

Whit his fix a non BGP host is detected and bgpLocalAs will be NULLed
```
./discovery.php -h 269 -m bgp-peers -d
...
#### Load disco module bgp-peers ####
SNMP['/usr/bin/snmpget' '-v2c' '-c' 'COMMUNITY' '-OQUsv' '-m' 'BGP4-MIB' '-M' '/home/sgr/sync/code/librenms/mibs:/home/sgr/sync/code/librenms/mibs/cisco' 'udp:HOSTNAME:161' 'bgpLocalAs.0']
No Such Object available on this agent at this OID  
  
No BGP on hostSQL[UPDATE `devices` set `bgpLocalAs`=NULL WHERE device_id=? [269] 18.72ms] 
```
tcpdump
```
.... GetRequest(28)  15.2.0
.... GetResponse(28)  15.2.0=[noSuchObject]

```

This fixes #14752 or  Point 1 in https://github.com/librenms/librenms/issues/14598#issuecomment-1362033906 so we don't query the peering DB with nonexisting ASNs anymore

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [x] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply 14948`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by Bierchermuesli** [comment]
Patch rollet here, I will give an update tomorrow how it goes. current situation

```
select DISTINCT `hardware`,`bgpLocalAs` FROM `devices` WHERE `disabled` = 0 AND `ignore` = 0 AND bgpLocalAs < 2000
    -> ;
+-----------------------------------------+------------+
| hardware                                | bgpLocalAs |
+-----------------------------------------+------------+
| NULL                                    |          3 |
| C9300-24P                               |          0 |
| ASR-920-24SZ-M                          |          0 |
| C1117-4P                                |          0 |
| NULL                                    |         41 |
| C897VA-K9                               |          0 |
| N9K-C9336C-FX2                          |          1 |
| WS-C3560-8PC-S                          |          0 |
| N9K-C93180YC-EX                         |          1 |
| FastIron ICX 6430 48-port PoE+          |          1 |
| C9300-24T                               |          0 |
| NULL                                    |        131 |
| N9K-C9372PX-E                           |          1 |
| FWS624G                                 |          1 |
| WS-C3850-24T-E                          |          0 |
| FES 12GCFRouter                         |          1 |
| FWS648G                                 |          1 |
| NULL                                    |          1 |
| FastIron ICX 6430 24-port PoE+          |          1 |
| CISCO892F-K9                            |          0 |
| WS-C3650-24PS-L                         |         80 |
| Juniper ACX710 Internet Backbone Router |        508 |
| Ruckus ICX 7150 48-port POE+ Switch     |          1 |
| C892FSP-K9                              |          0 |
| NULL                                    |        197 |
| FESX424-PREM                            |          1 |
| Nexus5548 Chassis                       |          1 |
| Juniper EX2300-24T Switch               |        514 |
| Juniper EX2300-24T Switch               |        508 |
| WS-C3650-48TQ-E                         |          0 |
| WS-C3560CX-12TC-S                       |          0 |
+-----------------------------------------+------------+
```

**Comment by Bierchermuesli** [comment]
my ~150 wrong ASns are gone...  🥳 

regarding the failed CI task i need a second opinion: is there a chance of wrong test data vrp_ce12804-withvrf ? @PipoCanaja 
```
$ lnms dev:simulate
$ snmpwalk -v 2c -c vrp_ce12804-withvrf 127.1.6.1:1161  '-m' 'BGP4-MIB' '-M' './mibs:./mibs/huawei' bgpLocalAs -On
.1.3.6.1.2.1.15.2 = No Such Instance currently exists at this OID
```

I assume `bgpLocalAs: 1` in [tests/data/vrp_ce12804-withvrf.json](https://github.com/librenms/librenms/blob/master/tests/data/vrp_ce12804-withvrf.json#L24416) is wrong. as it's also not listed in [tests/snmpsim/vrp_ce12804-withvrf.snmprec](https://github.com/librenms/librenms/blob/master/tests/snmpsim/vrp_ce12804-withvrf.snmprec)

Compared to another huawei (maybe also a bit wrong?)
```
$ snmpwalk -v 2c -c vrp_ne8000 127.1.6.1:1161  '-m' 'BGP4-MIB' '-M' './mibs:./mibs/huawei' bgpLocalAs -On
.1.3.6.1.2.1.15.2.0 = Wrong Type (should be INTEGER): Gauge32: 26479
```

Compared to a this one (which is pretty sure right...):
```
$ snmpwalk -v 2c -c iosxr_asr9001 127.1.6.1:1161  '-m' 'BGP4-MIB' '-M' './mibs:./mibs/cisco' bgpLocalAs -On
.1.3.6.1.2.1.15.2.0 = INTEGER: 65056
```

**Comment by PipoCanaja** [comment]
> my ~150 wrong ASns are gone... 🥳
> 
> regarding the failed CI task i need a second opinion: is there a chance of wrong test data in `vrp_ce12804-withvrf` ? @PipoCanaja (contributor)
> 
> ```
> $ lnms dev:simulate
> $ snmpwalk -v 2c -c vrp_ce12804-withvrf 127.1.6.1:1161  '-m' 'BGP4-MIB' '-M' './mibs:./mibs/huawei' bgpLocalAs -On
> .1.3.6.1.2.1.15.2 = No Such Instance currently exists at this OID
> ```
> 
> I assume `bgpLocalAs: 1` in [tests/data/vrp_ce12804-withvrf.json](https://github.com/librenms/librenms/blob/master/tests/data/vrp_ce12804-withvrf.json#L24416) is wrong. as it's also not listed in [tests/snmpsim/vrp_ce12804-withvrf.snmprec](https://github.com/librenms/librenms/blob/master/tests/snmpsim/vrp_ce12804-withvrf.snmprec)
> 

Wrong is may be a bit excessive 😄 Let say "incomplete" and "forged" 😄 

But this is not a big issue. You just need to update the JSON files with the current snmprec data, and they will appear here in the Pull Request. This will help the reviewer to evaluate your code, and all tests must pass before the PR can be merged anyway.

**Comment by Bierchermuesli** [comment]
hmm. can you elaborate my(?) ToDo a bit?
.1.3.6.1.2.1.15.2  does not exist in the [vrp_ce12804-withvrf.snmprec](https://github.com/librenms/librenms/blob/master/tests/snmpsim/vrp_ce12804-withvrf.snmprec) at all. so i'm confusing how it comes to that json data at the first place?

File: includes/discovery/bgp-peers.inc.php
Comment (murrant) [inline includes/discovery/bgp-peers.inc.php:15]: Bizarre, it goes back to the original code using a walk to fetch this value when it is just a scalar.  Probably test data will need to be corrected.

```suggestion
    $bgpLocalAs = \SnmpQuery::get('BGP4-MIB::bgpLocalAs.0')->value();
```

**Comment by murrant** [comment]
I wonder about this check... seems like some devices implement bgp without implementing .1.3.6.1.2.1.15.2

**Comment by Bierchermuesli** [comment]
So, how can me move on here?

[vrp_ce12804-withvrf.snmprec](https://github.com/librenms/librenms/blob/master/tests/snmpsim/vrp_ce12804-withvrf.snmprec) is abiously wrong. there is no standard `.1.3.6.1.2.1.15.2` nor the `1.3.6.1.4.1.2011.5.25.177.1.1.1.1` has any bgpLocalAs value. 

The  [1.3.6.1.4.1.2011.5.25.177.1.1.2.1.6 ](https://github.com/librenms/librenms/blob/759390ba4b72f2557e5213816f42f39d8ecd01ae/tests/snmpsim/vrp_ce12804-withvrf.snmprec#L3704C12-L3704C61) entry belongs to _The Counter That Records the Times the Remote BGP_ and not value 1 which is expected

I can implement an exeption but 'null' is in my opinion the correct value for this device.

**Comment by PipoCanaja** [comment]
Could be that [vrp_ce12804-withvrf.snmprec](https://github.com/librenms/librenms/blob/master/tests/snmpsim/vrp_ce12804-withvrf.snmprec) is incomplete. I can collect a new file, as I have one of these in production, and I was the one who collected this file a (long) time ago. 
I'll update the PR with a newly collected file, with your PR applied. We'll see how it goes.

**Comment by PipoCanaja** [comment]
Ok. Not sure why it was written like getNext in the 1st place. Could it be that some funky devices are not replying to BGP4-MIB::bgpLocalAs.0 but BGP4-MIB::bgpLocalAs.1 for instance ? or any other index ? 
Anyway, your change follows the MIB and indeed the vrp-ce12804 data is crap. Cause it works well in real life on my device.

So I'll correct the snmprec file to make it fit.

**Comment by PipoCanaja** [comment]
File corrected. I checked all the VRP test data on my dev machine. CI will test the others and return the result here. If tests are OK, then this PR will make it tomorrow.

**Review by PipoCanaja — APPROVED** [review]
Let's go

**Pull request merged**
